presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -991,6 +992,15 @@ protected Scope visitExplain(Explain node, Optional<Scope> scope)
         @Override
         protected Scope visitQuery(Query node, Optional<Scope> scope)
         {
+            Statement statement = analysis.getStatement();
+
+            if (statement instanceof Query && analysis.isStatementRewriteWithOptimizationEnabled() && SystemSessionProperties.isQueryOptimizationWithMaterializedViewEnabled(session)) {
+                Optional<Query> optimizeQuery = optimizeQueryUsingMaterializedView(node);
+                if (optimizeQuery.isPresent()) {
+                    return process(optimizeQuery.get(), scope);
```

Comment:
Hi @jainxrohit , when we call process(optimizeQuery) here, it will call visitTable() on the MV table, since optimizeQuery is on the MV table. Then current visitTable() implementation will do partition stitching and expand Q(mv) => Q(mv) union Q(baseTable), as we do for SELECT FROM mv.

My idea is to use the isStatementRewriteWithOptimizationEnabled() flag and add it to the condition of partition stitching in visitTable(). What do you think?

Reply:
The idea here is
Q(BaseTable) => Q(MV) => Q(MV) Union Q(BaseTable)